filters/include/pcl/filters/experimental/grid_filter.h

```diff
@@ -0,0 +1,234 @@
+   */
+  inline void
+  getFilterLimits(double& limit_min, double& limit_max) const
+  {
+    limit_min = filter_limit_min_;
+    limit_max = filter_limit_max_;
+  }
+
+  /** \brief Set to true if we want to return the data outside the interval specified by
+   * setFilterLimits (min, max). Default: false. \param[in] limit_negative return data
+   * inside the interval (false) or outside (true)
+   */
+  inline void
+  setFilterLimitsNegative(const bool limit_negative)
+  {
+    filter_limit_negative_ = limit_negative;
+  }
+
+  /** \brief Get whether the data outside the interval (min/max) is to be returned
+   * (true) or inside (false). \param[out] limit_negative true if data \b outside the
+   * interval [min; max] is to be returned, false otherwise
+   */
+  inline void
+  getFilterLimitsNegative(bool& limit_negative) const
+  {
+    limit_negative = filter_limit_negative_;
+  }
+
+  /** \brief Get whether the data outside the interval (min/max) is to be returned
+   * (true) or inside (false). \return true if data \b outside the interval [min; max]
+   * is to be returned, false otherwise
+   */
+  inline bool
+  getFilterLimitsNegative() const
+  {
+    return filter_limit_negative_;
+  }
+
+protected:
+  /** \brief Set to true if all fields need to be downsampled, or false if just XYZ. */
+  bool downsample_all_data_;
+
+  /** \brief Set to true if leaf layout information needs to be saved in \a
+   * leaf_layout_. */
+  bool save_leaf_layout_;
+
+  /** \brief The desired user filter field name. */
+  std::string filter_field_name_;
+
+  /** \brief The minimum allowed filter value a point will be considered from. */
+  double filter_limit_min_;
+
+  /** \brief The maximum allowed filter value a point will be considered from. */
+  double filter_limit_max_;
+
+  /** \brief Set to true if we want to return the data outside (\a filter_limit_min_;\a
+   * filter_limit_max_). Default: false. */
+  bool filter_limit_negative_;
+
+  /** \brief Minimum number of points per voxel for the centroid to be computed */
+  unsigned int min_points_per_voxel_;
+
+  /** \brief The iterable grid object for storing information of each fraction of space
+   * in the filtering space defined by the grid */
+  Grid grid_;
+
+  /** \brief Downsample a Point Cloud using a voxelized grid approach
+   * \param[out] output the resultant point cloud message
+   */
+  void
+  applyFilter(PointCloud& output) override
+  {
+    // Has the input dataset been set already?
+    if (!input_) {
+      PCL_WARN("[pcl::%s::applyFilter] No input dataset given!\n",
+               getClassName().c_str());
+      output.width = output.height = 0;
+      output.clear();
+      return;
+    }
+
+    // Copy the header (and thus the frame_id) + allocate enough space for points
+    output.height = 1;      // downsampling breaks the organized structure
+    output.is_dense = true; // we filter out invalid points
+
+    if (!GridStruct::setUp(this, grid_)) {
```

Comment:
At least for `VoxelGrid`, it can be failed. The range of `float` is larger than `size_t`, otherwise ["noise effect"](https://github.com/PointCloudLibrary/pcl/issues/4365#issuecomment-687721827) will happen due to unsigned int "wrap-around"

Reply:
If setup fails, we should return an empty cloud or the original cloud?

Reply:
It returns the original cloud `input_` now, does current way look good?